### ISSUE TITLE
No files found for pattern(s)

### ISSUE THREAD

**Issue by dfenerski** [SAP/grunt-openui5#171]
Hi guys, 

I am having a bit of trouble generating the Component-Preload file.
In my Gruntfile I have the following configuration:

```
openui5_preload: {
      board: {
        options: {
          resources: {
            cwd: "src/board",
            expand: true,
            src: [
              "model/*.json",
              "view/*.view.xml",
              "fragment/*.fragment.xml",
              "controller/*.controller.js",
              "util/*.js",
              "lib/classes/*.js",
              "lib/controls/*.js",
              "Component.js",
            ],
          },
          dest: "dest/board",
          compress: true,
        },
        components: true,
        filter: function (sPath) {
          return !(
            sPath.includes("lib/lib1") ||
            sPath.includes("lib/lib2") ||
            sPath.includes("lib/lib3") ||
            sPath.includes("sw.js")
          );
        },
      },
    },
```

And my projects structure is as follows:
![image](https://user-images.githubusercontent.com/48520530/95332434-b9871680-08ab-11eb-900f-3fdca38c4ae6.png)

When I run grunt openui5_preload:board I am getting the warning Warning: No files found for pattern(s): "./**", "!./Component-preload.js"! Use --force to continue.

What am I doing wrong?

Any help will will be deeply appreciated.

**Comment by matz3** [comment]
Could you please share the log output with the `--verbose` flag?
Please also make sure to use the latest `grunt-openui5` and `grunt` versions. Is this something that has worked before?

**Comment by dfenerski** [comment]
We never had something like this before, doing it for the first time.
This is the log output with the --verbose parameter.
`$ grunt openui5_preload:board --verbose
Initializing
Command-line options: --verbose, --gruntfile=C:\Users\Dimitar Fenerski\Project1\frontend\Gruntfile.js

Reading "Gruntfile.js" Gruntfile...OK

Registering Gruntfile tasks.
Initializing config...OK

Registering "grunt-exec" local Npm module tasks.
Reading C:\Users\Dimitar Fenerski\Project1\frontend\node_modules\grunt-exec\package.json...OK
Parsing C:\Users\Dimitar Fenerski\Project1\frontend\node_modules\grunt-exec\package.json...OK
Loading "exec.js" tasks...OK
+ exec

Registering "grunt-openui5" local Npm module tasks.
Reading C:\Users\Dimitar Fenerski\Project1\frontend\node_modules\grunt-openui5\package.json...OK
Parsing C:\Users\Dimitar Fenerski\Project1\frontend\node_modules\grunt-openui5\package.json...OK
Loading "connect.js" tasks...OK
+ openui5_connect
Loading "preload.js" tasks...OK
+ openui5_preload
Loading "theme.js" tasks...OK
+ openui5_theme
Loading "Gruntfile.js" tasks...OK
+ default

Running tasks: openui5_preload:board

Running "openui5_preload:board" (openui5_preload) task
Verifying property openui5_preload.board exists in config...OK
File: [no files]
Options: resources={"cwd":"src/board","expand":true,"src":["model/*.json","view/*.view.xml","fragment/*.fragment.xml","controller/*.controller.js","util/*.js","lib/classes/*.js","lib/controls/*.js","Component.js"]}, dest="dest/board", compress, compatVersion="edge"

Collecting resources
resource: cwd="src/board", expand, src=["model/*.json","view/*.view.xml","fragment/*.fragment.xml","controller/*.controller.js","util/*.js","lib/classes/*.js","lib/controls/*.js","Component.js"], prefix=""
Collecting model/local.json (src\board\model\local.json)...OK
Collecting view/Home.view.xml (src\board\view\Home.view.xml)...OK
Collecting view/Root.view.xml (src\board\view\Root.view.xml)...OK
Collecting fragment/Fragment1.fragment.xml (src\board\fragment\Fragment1.fragment.xml)...OK
Collecting fragment/Fragment2.fragment.xml (src\board\fragment\Fragment2.fragment.xml)...OK
Collecting fragment/Fragment3.fragment.xml (src\board\fragment\Fragment3.fragment.xml)...OK
Collecting controller/Home.controller.js (src\board\controller\Home.controller.js)...OK
Collecting controller/Root.controller.js (src\board\controller\Root.controller.js)...OK
Collecting util/dataFunctions.js (src\board\util\dataFunctions.js)...OK
Collecting util/formatter.js (src\board\util\formatter.js)...OK
Collecting util/options.js (src\board\util\options.js)...OK
Collecting util/sqlBuilder.js (src\board\util\sqlBuilder.js)...OK
Collecting lib/classes/Graph.js (src\board\lib\classes\Graph.js)...OK
Collecting lib/controls/Control1.js (src\board\lib\controls\Control1.js)...OK
Collecting lib/controls/Control2.js (src\board\lib\controls\Control2.js)...OK
Collecting lib/controls/Control3.js (src\board\lib\controls\Control3.js)...OK
Collecting Component.js (src\board\Component.js)...OK

Creating preload module for Component.js
Warning: No files found for pattern(s): "./**", "!./Component-preload.js"! Use --force to continue.

Aborted due to warnings.`

I also tried using the ui5 cli for building but had even less success there because of our multi-ui5 + express structure.

Thanks for the support

**Comment by matz3** [comment]
You need to provide a [`prefix`](https://github.com/SAP/grunt-openui5#prefix) with the component namespace as defined in your Component.js / manifest.json.

**Comment by dfenerski** [comment]
Thank you for the help. This was indeed something that I was missing. However it still fails. This is the output using --verbose now.

_Initializing
Command-line options: --verbose, --gruntfile=C:\Users\Dimitar Fenerski\Projects\Project1\frontend\Gruntfile.js

Reading "Gruntfile.js" Gruntfile...OK

Registering Gruntfile tasks.
Initializing config...OK

Registering "grunt-exec" local Npm module tasks.
Reading C:\Users\Dimitar Fenerski\Projects\Project1\frontend\node_modules\grunt-exec\package.json...OK
Parsing C:\Users\Dimitar Fenerski\Projects\Project1\frontend\node_modules\grunt-exec\package.json...OK
Loading "exec.js" tasks...OK
+ exec

Registering "grunt-openui5" local Npm module tasks.
Reading C:\Users\Dimitar Fenerski\Projects\Project1\frontend\node_modules\grunt-openui5\package.json...OK
Parsing C:\Users\Dimitar Fenerski\Projects\Project1\frontend\node_modules\grunt-openui5\package.json...OK
Loading "connect.js" tasks...OK
+ openui5_connect
Loading "preload.js" tasks...OK
+ openui5_preload
Loading "theme.js" tasks...OK
+ openui5_theme
Loading "Gruntfile.js" tasks...OK
+ default

Running tasks: openui5_preload:board

Running "openui5_preload:board" (openui5_preload) task
Verifying property openui5_preload.board exists in config...OK
File: [no files]
Options: resources={"cwd":"src/board","expand":true,"prefix":"Project1","src":["model/*.json","view/*.view.xml","fragment/*.fragment.xml","controller/*.controller.js","util/*.js","lib/classes/*.js","lib/controls/*.js","Component.js"]}, dest="dest/board", compress, compatVersion="edge"

Collecting resources
resource: cwd="src/board", expand, prefix="Project1", src=["model/*.json","view/*.view.xml","fragment/*.fragment.xml","controller/*.controller.js","util/*.js","lib/classes/*.js","lib/controls/*.js","Component.js"]
Collecting Project1/model/local.json (src\board\model\local.json)...OK
Collecting Project1/view/Home.view.xml (src\board\view\Home.view.xml)...OK
Collecting Project1/view/Root.view.xml (src\board\view\Root.view.xml)...OK
Collecting Project1/Component.js (src\board\Component.js)...OK

Creating preload module for Project1/Component.js
Reading src\board\model\local.json...OK
Adding Project1/model/local.json (1.21 kB → 847 B)
Reading src\board\view\Home.view.xml...OK
Adding Project1/view/Home.view.xml (32.7 kB → 23.82 kB)
Reading src\board\view\Root.view.xml...OK
Adding Project1/view/Root.view.xml (216 B → 173 B)
Reading src\board\controller\Home.controller.js...OK
Warning: Cannot read property 'length' of undefined Use --force to continue.

Aborted due to warnings.
_

There are no syntax or runtime errors in the specified files. When I try using --force, I still don't get any result. Do you have any clues? Thanks in advance...

**Comment by matz3** [comment]
Can you also run it with `--stack` to see the stack trace of the error?

**Comment by dfenerski** [comment]
```
TypeError: Cannot read property 'length' of undefined
    at C:\Users\Dimitar Fenerski\Projects\Project1\node_modules\grunt-openui5\tasks\preload.js:281:38
    at Array.forEach (<anonymous>)
    at C:\Users\Dimitar Fenerski\Projects\Project1\node_modules\grunt-openui5\tasks\preload.js:226:19
    at Array.forEach (<anonymous>)
    at C:\Users\Dimitar Fenerski\Projects\Project1\node_modules\grunt-openui5\tasks\preload.js:190:18
    at Array.forEach (<anonymous>)
    at C:\Users\Dimitar Fenerski\Projects\Project1\node_modules\grunt-openui5\tasks\preload.js:181:22
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (C:\Users\Dimitar Fenerski\Projects\Project1\node_modules\grunt-openui5\tasks\preload.js:108:31)
    at Object.<anonymous> (C:\Users\Dimitar Fenerski\Projects\Project1\node_modules\grunt\lib\grunt\task.js:252:15)

```

**Comment by matz3** [comment]
Okay, so maybe an issue with terser. What's the content of `src\board\controller\Home.controller.js` ?
And which exact `grunt-openui5` and `terser` version do you have installed?
Please run `npm ls grunt-openui5` and `npm ls terser`.

**Comment by dfenerski** [comment]
```
$ npm ls grunt-openui5
smartboard@1.0.0 C:\Users\Dimitar Fenerski\Projects\Project1
`-- grunt-openui5@0.17.0 
```
```
$ npm ls terser
smartboard@1.0.0 C:\Users\Dimitar Fenerski\Projects\Project1
`-- grunt-openui5@0.17.0
  `-- terser@4.8.0 

```

`Home.controller.js` is about 2500 lines and I am not entirely sure how much of it I can share. There are no syntax errors in it, however.

**Comment by matz3** [comment]
I suspect that it includes some new ES features that are supported by your browser but not by terser. And the error handling doesn't seem to be working.

Recently we've upgraded to terser v5 which adds more support for newer ES features, but there isn't a release, yet. Could you please try the current master state?
`npm install -D SAP/grunt-openui5#master`

**Comment by dfenerski** [comment]
I can definitely do that. I will edit my comment with results. I make extensive use of some newer JS features, but as I am also using Babel to transpile them, my idea was to pass to the preloader the transpiled files. Won't that work?

**Comment by dfenerski** [comment]
It worked! Thank you for the help and support, Matthias. 

My thinking was that as I use Babel, the transpiled files would be the ones that go into the preloader. Won't that work? What is the correct of tasks in this case?

**Comment by matz3** [comment]
Glad to hear that it works!
I've just released [v0.18.0](https://github.com/SAP/grunt-openui5/releases/tag/v0.18.0), which includes the mentioned terser upgrade.

Babel or any other transpiler needs to run before the other bundling steps like creating the preload.
And the bundling must be executed with the transpilation output as its input/sources.
